[wip] Add `mirror: true` opt that flips webcam picture to act like mirror

arturi opened the pull request:
- Should we add a UI button to flip/mirror the image? Camera might be looking the other way, it could be a front facing camera.
- Decide on what to do with dimensions, if we scale the image to fill the Dashboard UI window, we’ll have to crop some parts, depending on the size if the screen. What about mobile?

kvz commented:
> ? Camera might be looking the other way, it could be a front facing camera.

Out of curiosity, is it actually possible to switch between cameras too? (in the future?)

nqst commented:
> Should we add a UI button to flip/mirror the image?

I wouldn't add it, and would do it like on phones and in Photo Booth.app. So I think we can set `mirror: true` option by default, and allow users to change it (with a hint describing the cases when it's really needed). We also have to remember that once shot is made or video is recorded, the result must not be mirrored as in the preview. With that button, people might be more confused about how the result will look like.

BTW I couldn't test which camera is used on mobile when trying Webcam plugin — Safari asked me for the access, I allowed it, but nothing has changed, and the preview hasn't appeared. I expected to see the front camera by default.

If we're allowed to switch cameras (that's the case applicable only on mobile), and will make it one day, I think we should follow the same pattern as on mobile OSs — mirrored image for the front cam, normal image for the main cam. And of course, the switch camera button.

> Decide on what to do with dimensions, if we scale the image to fill the Dashboard UI window, we’ll have to crop some parts, depending on the size if the screen. What about mobile?

On desktop we can go without cropping and set the 100% width and automatic height based on proportions. Uppy window/frame can change height then, but I think it's fine. On mobile, IMO, the more we try to copy mobile OS' patterns, the more friendly UI will be.


What do you think?

arturi commented:
> BTW I couldn't test which camera is used on mobile when trying Webcam plugin — Safari asked me for the access, I allowed it, but nothing has changed, and the preview hasn't appeared. I expected to see the front camera by default.

😱I think it used to be disabled in Safari, because it did not support webcams. I will try to figure out why it’s not working, thanks for reporting.

> We also have to remember that once shot is made or video is recorded, the result must not be mirrored as in the preview. With that button, people might be more confused about how the result will look like.

Yeah, I spent some time implementing flipping on the result image as well, was sure that’s how it supposed to be, until I took a selfie with iPhone camera and it didn’t flip the result image 😱

> On desktop we can go without cropping and set the 100% width and automatic height based on proportions. Uppy window/frame can change height then, but I think it's fine. On mobile, IMO, the more we try to copy mobile OS' patterns, the more friendly UI will be.

This is the current version on screen with small height:

<img width="1273" alt="screen shot 2018-02-02 at 3 59 46 pm" src="https://user-images.githubusercontent.com/1199054/35754715-75336136-0832-11e8-95db-a0f154b8e296.png">

This is the full width version:

<img width="1273" alt="screen shot 2018-02-02 at 3 59 34 pm" src="https://user-images.githubusercontent.com/1199054/35754707-6f947b52-0832-11e8-81c0-51401c28c492.png">

I guess we’ll be needing a `uppy-Dashboard--tall` property to accompany `uppy-Dashboard--wide` we already have. But even then, it crops some parts of the image, as far as I can tell. I will push changes here with “full width” image on desktops, so you try if you’d like.

arturi commented:
I’ve found there is a way to choose between front and back cameras on mobile: https://developer.mozilla.org/en-US/docs/Web/API/MediaTrackConstraints/facingMode. Set to front as default, added it as Webcam plugin option. 

Set flipping for previews only, resulting image stays as is, like with mobile selfie cameras.

Fixed Webcam is Safari on iOS and Mac.

arturi commented:
@nqst do you think it might be problematic if we place the capture button in the lower center area it might cover the face on some screens? Maybe we should make it opaque or smaller 🤔

nqst commented:
> Do you think it might be problematic if we place the capture button in the lower center area it might cover the face on some screens? Maybe we should make it opaque or smaller 🤔

I believe it shouldn't be a problem, and also with the buttons at the bottom, we may slightly hint people to center face or other objects and not "stick" them to the edges.

Example — Instagram stories:

![artboard](https://user-images.githubusercontent.com/375537/36024038-228eb77c-0d97-11e8-89c4-c6b1e5fa3306.png)

But I also want to create a "safe" version with buttons on the bar (which is not overlaying the preview). We can choose what's best then.

ventoh commented:
I'm using uppy version 0.22.5 with facingMode: 'environment' as option in webcam plugin, but still no back camera access. Always front camera access. ? I'm on Android 5.1.0 Chrome 64.
Any help?

Thank you.

arturi commented:
@ventoh it’s out in the latest `0.23.0`: https://www.npmjs.com/package/uppy

goto-bus-stop commented on website/src/docs/webcam.md:
```diff
@@ -52,6 +55,20 @@ The types of recording modes to allow.
+- `user`: The video source is facing toward the user; this includes, for example, the front-facing camera on a smartphone.
+- `environment`:  The video source is facing away from the user, thereby viewing their environment. This is the back camera on a smartphone.
+- `left`: The video source is facing toward the user but to their left, such as a camera aimed toward the user but over their left shoulder.
+- `right`: The video source is facing toward the user but to their right, such as a camera aimed toward the user but over their right shoulder.
```
sweet!

nqst commented:
Here are the two versions of button placement. What do you think?

![modal camera 2x](https://user-images.githubusercontent.com/375537/36306852-33b798f8-1322-11e8-888e-a466d0aeaf74.jpg)

![modal camera alt 2x](https://user-images.githubusercontent.com/375537/36306856-36515798-1322-11e8-9344-38d7907030a8.jpg)

dogrocker commented:
Can facingMode change from `user` to `environment` dynamically with some button ? (ex. user want to switch camera from front to back).
